pkg/gatewayserver/gatewayserver.go

```diff
@@ -897,16 +899,19 @@ func (gs *GatewayServer) handleUpstream(ctx context.Context, conn connectionEntr
 			if msg.Message.Payload == nil {
 				msg.Message.Payload = &ttnpb.Message{}
 				if err := lorawan.UnmarshalMessage(msg.Message.RawPayload, msg.Message.Payload); err != nil {
-					registerDropUplink(ctx, gtw, msg, "validation", err)
 					continue
 				}
 			}
+			registerReceiveUplink(ctx, gtw, msg, protocol)
```

Comment:
I would still register the receive only at the end, after everything has been dropped.